### PR TITLE
Lift ppx expansion errors as extension nodes to top level of file

### PR DESCRIPTION
**Pull request by ceastlund**
Changes `map_top_down` to collect errors explicitly. Like #357 but with fewer added/duplicated classes, and with errors mostly handled by a `With_errors` monad.

### CODE REVIEW

**Review by pitag-ha — COMMENTED** [review]
Thanks, I like the idea of lifting the errors and treating them at signature/structure level a lot, so either #357 or this PR are definitely a good improvement! As a side-note: doing so introduces one small inconsistency: ppxlib errors will be lifted, while embedded errors by context free rules will stay at the nodes they arise at. I don't think that's a problem though.

The main differences I see with #357 are:
-  the `map_top_down` class doesn't get duplicated. I think that's a good thing. #357 could very easily be adapted to do the same though, right?
-  the `Ppxlib.Ast_traverse.map_with_expansion_context` gets erased. Could also be adapted though. See my comment about that below concerning `bisect_ppx`.
-  error handling is factored through the new `With_errors ` monad. That seems very nice to me and makes things very well readable!

I haven't had a close look though. Are there more differences?

In any case, if you aren't in a strong hurry with #352, I think it would be cool to wait for @panglesd to come back from vacation to see what he thinks, both in general about this PR and compared to his #357.

I haven't fully reviewed yet, but I've already dropped some comments below.

File: dune-project [hunk 1]
```diff
@@ -15,7 +15,7 @@
 (package
  (name ppxlib)
  (depends
-  (ocaml (and (>= 4.04.1) (< 5.1.0)))
+  (ocaml (and (>= 4.06) (< 5.1.0)))
```
[the remark this answers is not in the document]
Comment (ceastlund) [reply dune-project:18]: I used `let open ... in` inside a class definition. If it's important to keep newer ppxlib compatible with older compilers we could handle it another way.

File: CHANGES.md [hunk 2]
```diff
@@ -1,6 +1,8 @@
 unreleased
 -------------------
 
+- Update context-free rules to collect expansion errors and propagate
+  them to top level without failing. (#358, @ceastlund)
```
[the remark this answers is not in the document]
Comment (ceastlund) [reply CHANGES.md:5]: True. Although down the road we could extend context-free rules to be able to use this mechanism, too. They just don't currently have a way to return both expanded code and errors separately.

File: traverse_builtins/ppxlib_traverse_builtins.ml [hunk 3]
```diff
@@ -161,6 +162,52 @@ class virtual ['res] lift =
         | x :: l -> self#constr "::" [ f x; self#list f l ]
   end
 
+class virtual ['ctx, 'res] lift_map_with_context =
+  object (self)
+    method virtual other : 'a. 'ctx -> 'a -> 'res
```
Comment (panglesd) [inline traverse_builtins/ppxlib_traverse_builtins.ml:167]: Maybe it makes sens to have a `lift_with_context` type in `T`, just to increase readability, for methods `other`, `record`, `constr`, `tuple`

File: dune-project (hunk 1 above)
[the remark this answers is not in the document]
Comment (panglesd) [reply dune-project:18]: I would be in favor of using a global open and keep the current compatibility (although I'm not personally attach to keeping compatibility with very old versions, the `let open ... in` do not justify it I think!)

**Review by panglesd — APPROVED** [review]
I think that using a `lift_map` rather than a `fold_map` is better (we do not need to pass the accumulator when calling methods).

It also leads with the infix syntax to a cleaner code than the `fold_map`, although I generally prefer the `let` syntax (unfortunately let-prunning is unavailable in old ocaml versions...)

I have small comments on the code or the names, but that is good for merge on my side!

File: CHANGES.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (ceastlund) [reply CHANGES.md]: reworded

File: dune-project (hunk 1 above)
[the remark this answers is not in the document]
Comment (ceastlund) [reply dune-project]: okay, restored 4.04.1 compatibility

File: src/driver.ml [hunk 4]
```diff
@@ -233,10 +233,16 @@ module Transform = struct
         in
         map#structure base_ctxt (List.concat [ attrs; header; st; footer ])
       in
+      let st =
+        (errors
+        |> List.map ~f:Location.Error.to_extension
+        |> List.map ~f:(Extension.Context.node_of_extension Structure_item))
```
[the remark this answers is not in the document]
Comment (ceastlund) [reply src/driver.ml]: removed node_of_extension

File: src/ast_traverse.ml [hunk 5]
```diff
@@ -90,30 +96,74 @@ let var_names_of =
 let ec_enter_module_opt ~loc name_opt ctxt =
   Expansion_context.Base.enter_module ~loc (module_name name_opt) ctxt
 
-class map_with_expansion_context =
+class lift_map_with_expansion_context_and_errors =
```
[the remark this answers is not in the document]
Comment (ceastlund) [reply src/ast_traverse.ml]: okay, renamed

File: src/ast_traverse.ml [hunk 6]
```diff
@@ -90,30 +96,74 @@ let var_names_of =
...
+
+    method record _ctx fields = List.concat_map fields ~f:snd
+    method constr _ctx _tag args = List.concat args
+    method tuple _ctx l = List.concat l
 
     method! expression ctxt
         { pexp_desc; pexp_loc; pexp_loc_stack; pexp_attributes } =
       let ctxt = Expansion_context.Base.enter_expr ctxt in
-      let pexp_desc =
+      let pexp_desc, desc_res =
```
[the remark this answers is not in the document]
Comment (ceastlund) [reply src/ast_traverse.ml]: good point, renamed

File: src/context_free.ml [hunk 7]
```diff
@@ -574,54 +570,49 @@ class map_top_down ?(expect_mismatch_handler = Expect_mismatch_handler.nop)
...
-      { pcsig_self; pcsig_fields }
+      self#core_type base_ctxt pcsig_self >>= fun pcsig_self ->
+      map_nodes EC.class_type_field class_type_field super#class_type_field
+        (fun x -> x.pctf_loc)
+        base_ctxt pcsig_fields
+      >>| fun pcsig_fields -> { pcsig_self; pcsig_fields }
 
     (* TODO: try to factorize #structure and #signature without meta-programming *)
     (*$*)
     method! structure base_ctxt st =
```
[the remark this answers is not in the document]
Comment (ceastlund) [reply src/context_free.ml:586]: left as-is, this seemed a complex enough change I'd rather leave it separate

File: traverse_builtins/ppxlib_traverse_builtins.ml (hunk 3 above)
Comment (ceastlund) [reply traverse_builtins/ppxlib_traverse_builtins.ml:167]: I don't think we actually get any mileage out of that. I believe those methods show up in four places. In two, they have context arguments; in two they don't. In two (cutting across the other axis) they're virtual; in two they're concrete. So I don't think there's any actual abstraction to be had.

**Pull request merged**